### PR TITLE
Swapped the order in which the engine.pak is searched

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
First the Project Cache Root folder is searched before falling back to
the Executable Directory

Removed the need for the engine.json and project.json in a project
release layout when a "Cache" directory exist at the root.
The project root uses the the first "Cache" directory it finds by scanning upwards as if fails to find a
project.json,
The engine root use the project root, if it fails to reconcile the
engine path using project.json "engine" key and the o3de_manifest.json
"engines_path" object.

depends on #4454 

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

### CODE REVIEW

**Comment by amznestebanpapp** [comment]
If the `engine.json` and `project.json` are no longer needed, can you remove the generation from `cmake/Projects.cmake` and `camke/Platform/Common/Install_common.cmake`?

File: Code/Framework/AzFramework/AzFramework/Archive/ArchiveVars.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+#include <AzCore/base.h>
+#include <AzFramework/Archive/ArchiveVars.h>
+
+namespace AZ::IO
+{
+    FileSearchPriority GetDefaultFileSearchPriority()
+    {
+#if defined(O3DE_DEFAULT_FILE_SEARCH_MODE)
```
Comment (amznestebanpapp) [inline Code/Framework/AzFramework/AzFramework/Archive/ArchiveVars.cpp:16]: where is this defined? I imagine you want to pass this from cmake and have it as a cache variable

**Review by amznestebanpapp — CHANGES_REQUESTED** [review]

File: Code/Framework/AzFramework/AzFramework/Archive/ArchiveVars.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzFramework/AzFramework/Archive/ArchiveVars.cpp:16]: Ah this change, shouldn't be in this PR

**Comment by lumberyard-employee-dm** [comment]
> 
> 
> If the `engine.json` and `project.json` are no longer needed, can you remove the generation from `cmake/Projects.cmake` and `cmake/Platform/Common/Install_common.cmake`?

Yes, I'll make that change.

File: Code/Framework/AzGameFramework/AzGameFramework/Application/GameApplication.cpp [hunk 2]
```diff
@@ -28,20 +28,22 @@ namespace AzGameFramework
...
-            enginePakPath.clear();
-            if (m_settingsRegistry->Get(enginePakPath.Native(), AZ::SettingsRegistryMergeUtils::FilePathKey_CacheRootFolder))
-            {
-                // fall back to checking Project Cache Root.
-                enginePakPath /= "engine.pak";
-                m_archive->OpenPack("@assets@", enginePakPath.Native());
-            }
+            // fall back to checking Project Cache Root.
+            enginePakPath /= "engine.pak";
+            enginePakOpened =m_archive->OpenPack("@projectproductassets@", enginePakPath.Native());
```
Comment (AMZN-koppersr) [inline Code/Framework/AzGameFramework/AzGameFramework/Application/GameApplication.cpp:41]: Nitpick: space needed between `=m_archive`.

**Review by AMZN-koppersr — APPROVED** [review]

**Review by amznestebanpapp — APPROVED** [review]

**Review by amzn-phist — APPROVED** [review]

File: cmake/Projects.cmake [hunk 3]
```diff
@@ -173,12 +173,7 @@ if("${CMAKE_INSTALL_CONFIG_NAME}" MATCHES "^([Rr][Ee][Ll][Ee][Aa][Ss][Ee])$")
             message(STATUS "${install_output_folder}/engine.pak generated")
         endif()
     endif()
-    file(WRITE ${install_output_folder}/project.json 
-"{
-    \"project_name\": \"@project_name@\"
-}")
-endif()
-]=])
+
```
Comment (amzn-phist) [inline cmake/Projects.cmake:176]: Needs a closing `]=])` here.

**Review by amzn-phist — CHANGES_REQUESTED** [review]

File: cmake/Projects.cmake (hunk 3 above)
Comment (lumberyard-employee-dm) [reply cmake/Projects.cmake:176]: That it does.

Comment (amzn-phist) [reply cmake/Projects.cmake:176]: Actually, I think it also needs that `endif()` as well.  I'm getting Flow control error with these changes.

Comment (amzn-phist) [reply cmake/Projects.cmake:178]: Ah, you got it.

Comment (lumberyard-employee-dm) [reply cmake/Projects.cmake:178]: It has been fixed now.
You also need the changes from #4454

**Review by amzn-phist — APPROVED** [review]

File: cmake/Projects.cmake (hunk 3 above)
Comment (amzn-phist) [reply cmake/Projects.cmake:178]: Right, I commented that part out for now.

File: Code/Framework/AzGameFramework/AzGameFramework/Application/GameApplication.cpp [hunk 4]
```diff
@@ -28,20 +28,22 @@ namespace AzGameFramework
...
-        AZ::IO::FixedMaxPath enginePakPath = AZ::Utils::GetExecutableDirectory();
-        enginePakPath /= "engine.pak";
-        if (!m_archive->OpenPack("@assets@", enginePakPath.Native()))
+        // Attempt to mount the engine pak to the project product asset alias
+        // Search Order:
+        // - Project Cache Root Directory
+        // - Executable Directory
+        bool enginePakOpened{};
+        AZ::IO::FixedMaxPath enginePakPath;
+        if (m_settingsRegistry->Get(enginePakPath.Native(), AZ::SettingsRegistryMergeUtils::FilePathKey_CacheRootFolder))
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzGameFramework/AzGameFramework/Application/GameApplication.cpp:37]: This is where the `SettingsRegistryMergeUtils::FindProjectRoot` changes come into play.
It attempts to check if there is a "Cache" directory at or above the executable directory and if so, it sets that as the directory as the ProjectPath, in which case the CacheRootFolder gets set to `FilePathKey_ProjectPath / "Cache" / <asset-type>`

**Pull request merged**